content/courses/ada-in-practice/chapters/idioms_for_protected_objects.rst

```diff
@@ -0,0 +1,1760 @@
+:ada:`Wait` for :ada:`B` will be able to return. Otherwise, that task will be
+blocked and cannot return from the call. Later another task will set
+:ada:`Signaled (B)` to :ada:`True`, and then the waiting task can be unblocked.
+
+Since an aggregate can also contain only one component if desired, we can use a
+single set of protected routines for waiting and signaling in the
+:ada:`Manager` protected type. We don't need one set of routines for waiting
+and signaling a single event, and another set of routines for waiting and
+signaling multiple events. Here then is the visible part:
+
+.. code-block:: ada
+
+    protected type Manager is
+
+       entry Wait
+         (Any_Of_These : Event_List;
+          Enabler      : out Event);
+
+       procedure Signal (All_Of_These : Event_List);
+
+    private
+       ...
+    end Manager;
+
+Both the entry and the procedure take an argument of the array type, indicating
+one or more client events. The entry, called by waiting tasks, also has an
+output argument, :ada:`Enabler`, indicating which specific entry enabled the
+task to resume, i.e., which event was found signaled and was selected to
+unblock the task. We need that parameter because the task may have specified
+that any one of several events would suffice, and more than one could have been
+signaled.
+
+The bodies of our API routines are then just calls into the protected
+:ada:`Manager` that is passed as the first argument. For example, here are two
+of the four:
+
+.. code-block:: ada
+
+    procedure Wait
+      (This         : in out Manager;
+       Any_Of_These :        Event_List;
+       Enabler      :    out Event)
+    is
+    begin
+       This.Wait (Any_Of_These, Enabler);
+    end Wait;
+
+    procedure Signal
+      (This     : in out Manager;
+       This_One : Event)
+    is
+    begin
+       This.Signal (Event_List'(1 => This_One));
+    end Signal;
+
+Now let's examine the implementation of the protected type. The visible part is
+repeated here:
+
+.. code-block:: ada
+
+    protected type Manager is
+
+       entry Wait
+         (Any_Of_These : Event_List;
+          Enabler      : out Event);
+
+       procedure Signal (All_Of_These : Event_List);
+
+    private
+       ...
+    end Manager;
+
+The entry :ada:`Wait` suspends callers until one of the requested events is
+signaled, as specified by the argument :ada:`Any_Of_These`. Normally we'd
+expect to use the entry barrier to express this behavior by querying the
+events' state array. If one of the requested events is :ada:`True` the barrier
+would allow the call to execute and complete. However, barriers do not have
+compile-time visibility to the entry parameters, so the parameters cannot be
+referenced in the barriers. This situation calls for a requeue statement.
+
+Because :ada:`Wait` always takes a call, the entry barrier is just hard-coded
+to :ada:`True`. (That's always a strong indication that requeue is involved.)
```

Comment:
Let's change "(That's always a strong" to be "(As mentioned earlier, that's always a strong"